### PR TITLE
crypto.generateKeyPair('x25519') and crypto.diffieHellman({ privateKey, publicKey }) support

### PR DESCRIPTION
**Pull request by fuzzc0re**
Part of #45614

- [x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [x ] Test the change in your own code. (Compile and run.)

If changing an existing definition:

- [x ] Provide a URL to documentation or source code which provides context for the suggested changes:
    https://nodejs.org/api/crypto.html#crypto_crypto_generatekeypair_type_options_callback
    https://nodejs.org/api/crypto.html#crypto_crypto_diffiehellman_options

### CODE REVIEW

File: types/node/ts3.1/crypto.d.ts [hunk 1]
```diff
@@ -545,6 +551,16 @@ declare module "crypto" {
         };
     }
 
+    interface X25519KeyPairOptions<PubF extends KeyFormat, PrivF extends KeyFormat> {
+        publicKeyEncoding: {
+            type: 'pkcs1' | 'spki';
+            format: PubF;
+        };
+        privateKeyEncoding: BasePrivateKeyEncodingOptions<PrivF> & {
+            type: 'sec1' | 'pkcs8';
```
Comment (Flarna) [inline types/node/ts3.1/crypto.d.ts:560]: docs tell that `sec1` is only for EC. Is it also allowed for X25519?

**Review by Flarna — CHANGES_REQUESTED** [review]
Please add also some test.
These crypto APIs are hard to follow for non crypto people like me. But I expect that you have working code somewhere you can port into a test here.

File: types/node/ts3.1/crypto.d.ts [hunk 2]
```diff
@@ -545,6 +551,16 @@ declare module "crypto" {
         };
     }
 
+    interface X25519KeyPairOptions<PubF extends KeyFormat, PrivF extends KeyFormat> {
+        publicKeyEncoding: {
+            type: 'pkcs1' | 'spki';
```
[the remark this answers is not in the document]
Comment (fuzzc0re) [reply types/node/ts3.1/crypto.d.ts]: Sorry for that I just copied the committed ed25519 interface and changed letters. I corrected both in ac52918.

File: types/node/ts3.1/crypto.d.ts (hunk 1 above)
Comment (fuzzc0re) [reply types/node/ts3.1/crypto.d.ts]: Same as above. Sorry again.

**Comment by fuzzc0re** [comment]
The test in aa846a5 does not demonstrate a full symmetric key example with kdf and salt. It just tests that crypto.diffieHellman generates shared secrets properly.

**Review by Flarna — APPROVED** [review]

**Comment by fuzzc0re** [comment]
Adds Typescript support for https://github.com/nodejs/node/issues/26626#issuecomment-576718196.

**Review by Flarna — APPROVED** [review]

**Comment by orta** [comment]
OK, looks good to me too - I'll leave it to a contributor to give it one more look over and you're welcome to merge

**Comment by fuzzc0re** [comment]
Ready to merge

**Comment by yoursunny** [comment]
Version 14.6.4 is *sometimes* causing this error:
```
error TS2688: Cannot find type definition file for 'node'.
```
https://travis-ci.com/github/yoursunny/NDNts/builds/182865199

The CI build randomly fails with the above error on some jobs, but if I retrigger the build, it would fail on a different set of jobs. However, I don't get this error if I downgrade to @types/node 14.6.3.

I have a monorepo with pnpm and other weird stuff, so it's entirely possible that I messed up somewhere.
If anyone else is getting the same error, we can investigate more.

**Comment by fuzzc0re** [comment]
@yoursunny do you still get the same error? I am not sure but could it be the case that crypto.diffieHellman({ }) was added in node v13.9.0 and your supporting v12 is causing the error? I am not familiar with your setup so I might be completely wrong.

**Comment by yoursunny** [comment]
> crypto.diffieHellman({ }) was added in node v13.9.0

I don't think it has anything to do with the *contents* of this PR. It's more likely that *the way it's packaged and published* was causing the problem.
I downgraded to 14.6.3 for now.
